Changes for predicate gas metering (#448)
FuelLabs/fuel-specs: src/vm/index.md

## Contexts

There are 4 _contexts_ in the FuelVM: [predicate estimation](#predicate-estimation), [predicate verification](#predicate-verification), [scripts](#script-execution), and [calls](./instruction_set.md#call-call-contract). A context is an isolated execution environment with defined [memory ownership](#ownership) and can be _external_ or _internal_:

- External: predicate and script. `$fp` will be zero.
- Internal: call. `$fp` will be non-zero.

[Returning](./instruction_set.md#return-return-from-call) from a context behaves differently depending on whether the context is external or internal.

## Predicate Estimation

For any input of type [`InputType.Coin`](../protocol/tx_format/index.md), a non-zero `predicateLength` field means the UTXO being spent is a [P2SH](https://en.bitcoinwiki.org/wiki/P2SH) rather than a [P2PKH](https://en.bitcoinwiki.org/wiki/Pay-to-Pubkey_Hash) output.

For each such input in the transaction, the VM is [initialized](#vm-initialization), then:

1. `$pc` and `$is` are set to the start of the input's `predicate` field.
1. `$ggas` and `$cgas` are set to `tx.gasLimit`.

Predicate verification will fail if gas is exhausted during execution.

> After successful execution, `predicate.gasUsed` is set to `tx.gasLimit - $cgas`.

Reviewer: Even though we don't allow calls, I think it's safer to use ggas instead of cgas.

```suggestion
After successful execution, `predicate.gasUsed` is set to `tx.gasLimit - $ggas`.
```

## Predicate Verification

For any input of type [`InputType.Coin`](../protocol/tx_format/index.md), a non-zero `predicateLength` field means the UTXO being spent is a [P2SH](https://en.bitcoinwiki.org/wiki/P2SH) rather than a [P2PKH](https://en.bitcoinwiki.org/wiki/Pay-to-Pubkey_Hash) output.

For each such input in the transaction, the VM is [initialized](#vm-initialization), then:

1. `$pc` and `$is` are set to the start of the input's `predicate` field.
1. `$ggas` and `$cgas` are set to `tx.gasLimit`.

> Predicate verification will fail if gas is exhausted during execution. Additionally, if `predicate.gasUsed` is nonzero, a check will be performed to ensure that less than `predicate.gasUsed` gas is consumed:

Reviewer: We shouldn't include the non-zero logic here now that we have estimation.

```suggestion
Predicate verification will fail if gas is exhausted during execution. Additionally, a check will be performed to ensure that less than `predicate.gasUsed` gas is consumed:
```

```pseudo

> if $cgas < tx.gasLimit - predicate.gasUsed {
>     return false
> }

Reviewer: ```suggestion
# invert predicate.gasUsed by the tx.gasLimit to simplify comparison with $cgas
min_allowed_gas = tx.gasLimit - predicate.gasUsed
if $cgas < min_allowed_gas {
    return false
}
```

> ```

Reviewer: We should also have a check that the sum of all `predicate.gasUsed` <= `tx.gasLimit`

A predicate that halts without returning Boolean `true` would not pass verification, making the entire transaction invalid. Note that predicate validity is monotonic with respect to time (i.e. if a predicate evaluates to `true` then it will always evaluate to `true` in the future).

After successful execution, `predicateGasUsed` is set to `tx.gasLimit - $ggas`.

## Predicate Verification

For any input of type [`InputType.Coin`](../protocol/tx_format/input.md#inputcoin) or [`InputType.Message`](../protocol/tx_format/input.md#inputmessage), a non-zero `predicateLength` field means the UTXO being spent is a [P2SH](https://en.bitcoinwiki.org/wiki/P2SH) rather than a [P2PKH](https://en.bitcoinwiki.org/wiki/Pay-to-Pubkey_Hash) output.

For each such input in the transaction, the VM is [initialized](#vm-initialization), then:

1. `$pc` and `$is` are set to the start of the input's `predicate` field.
1. `$ggas` and `$cgas` are set to `predicateGasUsed`.

Predicate verification will fail if gas is exhausted during execution.

During predicate mode, hitting any of the following instructions causes predicate verification to halt, returning Boolean `false`:

> 1. Any [contract instruction](./instruction_set.md#contract-instructions).

Reviewer 2: Should this be reworded a bit since there is no list anymore - it's only one item. Maybe something like

> During predicate mode, hitting any [contract instruction](./instruction_set.md#contract-instructions) causes predicate verification to halt, returning Boolean `false`.

Author: I agree, I'll get this in now.

Author: done

For each such input in the transaction, the VM is [initialized](#vm-initialization), then:

1. `$pc` and `$is` are set to the start of the input's `predicate` field.

> 1. `$ggas` and `$cgas` are set to the lower of `tx.gasLimit` or the remaining gas following the previous predicate execution.

Reviewer 3: Hmmm. There's a few issues with this:

1. It creates a sequential dependence on lower-position predicates. Without any prior information on how much gas is used by each predicate, this requires predicates be verified sequentially.
2. Even _with_ prior information, this change makes predicates no longer position-independent. Which can make partially-signed transactions more complex to implement.

Overall, I'd prefer something like a fixed max gas for each predicate, independently. And this max gas can be a consensus rule.

If the gas schedule is changed, then predicates may become invalid. But that's the case even if we allow users to define the max gas since there's a hard cap in the form of the block gas limit.

Reviewer: Good points. Having a per predicate gas limit would be better for parallelization. Would there be both a per-predicate gasLimit specified + consensus rule max gas per predicate? Or would each predicate always assume it has the consensus based max gas, potentially locking up more funds than needed for execution?

The main benefits of using the tx-level gas limit as opposed to predicate level is:
1. UX simplicity - it is easier for the sdk to estimate the tx gas limit regardless of changes to the gas schedule
2. Smaller tx payload, as each new field increases our calldata expense / bandwidth
3. Users only need to lock up the minimal amount of necessary funds to pay for execution, potentially allowing for more user-side parallelism by minimizing gas lockup overhead (assuming there's no per-predicate user configurable gas limit)

Some ways to do parallel execution with this approach could be:

- Simply execute all predicates in parallel using the same global tx.gasLimit for each one. Total up the amount of gas each predicate consumed and fail the verification after execution if they ran over the limit. In the best case this doesn't add any overhead and keeps the predicate UX simpler. In the worst case the VM may spend more time executing instructions than it should.
   - The main risk with this approach is if the txpool would do extra work, it's possible it could be overloaded more easily. However, since it has to evaluate potentially invalid predicates anyways without being able to collect any fees, it's hard to say if this really worsens the situation for the txpool. 
   - The other risk would be during bft consensus. If a block proposed for the next round exceeded the allowed execution budget it could cause a lag in voting (although it would have to be some very large predicates at the very end of the block). However, given BFT only supports a small validator set, proposing a bad block which exceeds the execution budget could be a punishable/slashable offence which would deter this from happening.
- Use an atomic counter external to the cgas/ggas registers to track the total gas usage across multiple VM instances. However, this could add execution overhead as charging gas for predicates would require an additional CAS operation.

Reviewer: We had a call with @FuelLabs/client and @adlerjohn to discuss this in more detail -  summarizing the notes here:

- Sharing a global gasLimit prevents parallelization between predicates and script execution. We need to know how much gas predicates can use in advance, otherwise if there's an overage we would lose determinism as to whether the script status is reverted or if the predicates are invalid, thus invalidating the entire block.
- We will add a new malleable field to predicates called gasUsed. This avoids the need to have predicate amounts cover a global gas limit per predicate which could create a lot of dust.
- Each predicate will have cgas and ggas always initialized to the tx.gasLimit
- If predicate.gasUsed is set, the vm will ensure the current gas consumption doesn't exceed that (resolving potential DOS concerns)
- predicate.gasUsed will not be signed over, but the mempool will expect it to be set correctly by the user
- The user should be able to estimate the gas usage of each predicate via a single tx dry-run
- predicate.gasUsed is zeroed out on the tx during predicate verification

Reviewer 4: > The user should be able to estimate the gas usage of each predicate via a single tx dry-run

And we can execute the predicate without a `dry-run` on the SDK side potentially in the future(because we don't need access to the on-chain state)

For each such input in the transaction, the VM is [initialized](#vm-initialization), then:

> 1. `$pc` and `$is` are set to the start of the input's `predicate` field.

Reviewer: Maybe we should be more specific here around how gas factors into predicate verification? If we introduce the concept of `remaining_gas` here it would make the script execution update more clear.

Author: how should we ensure that we don't overly specify implementation for potentially parallelizing predicate execution? we eventually wouldn't be initializing predicate execution with remaining gas each time, just aggregating the resultant gas used.

Reviewer: Good point, although even if we don't include remaining gas in initialization it should still be explicitly defined before using it for script initialization. Something like:

```math
remaining_gas =  \sum_{i=0}^p gas(predicates[i])
```

The VM is [initialized](#vm-initialization), then:

1. `$pc` and `$is` are set to the start of the transaction's script bytecode.

> 1. `$ggas` and `$cgas` are set to `tx.gasLimit`. If predicate execution has decreased remaining gas, gas limit is initialized to remaining gas.

Reviewer: ```suggestion
1. `$ggas` and `$cgas` are set to lower of `tx.gasLimit` or `remaining_gas`. `remaining_gas` is determined by deducting any gas consumed during predicate execution from `tx.gasLimit`.
```

> A predicate that halts without returning Boolean `true` does not pass verification, making the entire transaction invalid. Note that predicate validity is monotonic with respect to time (i.e. if a predicate evaluates to `true` then it will always evaluate to `true` in the future).

Reviewer: > During predicate mode, hitting any of the following instructions causes predicate verification to halt, returning Boolean `false`:
> 1. Any [contract instruction](./instruction_set.md#contract-instructions).
> 1. [JMP](./instruction_set.md#jmp-jump), [JI](./instruction_set.md#ji-jump-immediate), [JNE](./instruction_set.md#jne-jump-if-not-equal), [JNEI](./instruction_set.md#jnei-jump-if-not-equal-immediate), or [JNZI](./instruction_set.md#jnzi-jump-if-not-zero-immediate) with jump-to value less than or equal to `$pc` (these would allow loops). In other words, `$pc` must be strictly increasing.

The section above this needs to adjust the restriction on JMP to allow backwards jumps within the predicate range
